### ISSUE TITLE
Not displaying days left in natural language after 2023.4

### ISSUE THREAD

**Issue by angarok** [DSorlov/swemail#19]
Prior to Home Assistant release 2023.4 the sensor displayed the time remaining to next delivery in natural language, e.g. "Tomorrow" or "In three days". After updating Home Assistant to 2023.4 (and since 2023.4.1) only the number of remaining days is shown, e.g "4".

**Comment by Maggie70** [comment]
Samma här

**Comment by DSorlov** [comment]
I am unable to replicate this on any versions. Logs or anything which can help?

**Comment by andjo** [comment]
Having the same issue on 2023.4.2. I see nothing related in the logs, but not sure I'm looing at the correct place.

Just taking a wild guess that it is related to the custom device class, is that a valid thing to do?
![image](https://user-images.githubusercontent.com/665563/230997522-d676271d-cd91-4d89-abe5-28378bc6fe0f.png)

**Comment by andjo** [comment]
Nope, just a bunch of `console.info()` from other addons loading.
![image](https://user-images.githubusercontent.com/665563/230999231-613dbcd8-044f-4a1b-b640-d238ecfe9a97.png)

**Comment by Steffe2** [comment]
Not that it really helps, but same issue here (2023.4.2).

Haven't done much troubleshooting in HA, so if there's anything you need log-wise, I probably need directions 😅

**Comment by angarok** [comment]
Have been taking a look at the code to try to figure out what stopped working. I'm not a coder so I can't really do much here. One observation is that the natural language strings is actually included in the custom component, not generated by HA Core. But I can't find where in the code the sensor maps days remaining (variable "numDays") to the array in the translation json? And where does it select the right json-file with translations depending on language settings?

**Comment by RH-Dreambox** [comment]
Samma problem här, siffran för antal dagar visas men den konverteras inte till texten i filen "sensor.sv.json".
Har installerat om applikation och integration men det hjälper inte.

**Comment by JohanAlvedal** [comment]
Samma problem här

**Comment by thulin82** [comment]
Same here, update from 2023.4.1 to 2023.4.2 broke the counter

**Comment by broberg88** [comment]
I also have problems, it comes back to today sometimes but after a while its back to numbers. 
Home Assistant 2023.4.4
Supervisor 2023.04.0
Operating System 9.5
Frontend 20230411.1 - latest

**Comment by NCC-1031** [comment]
Samma problem här.

**Comment by forzaKGB** [comment]
Same issue since upgrading to 2023.4 (currently at 2023.4.3)

**Comment by thulin82** [comment]
> Have been taking a look at the code to try to figure out what stopped working. I'm not a coder so I can't really do much here. One observation is that the natural language strings are actually included in the custom component, not generated by HA Core. But I can't find where in the code the sensor maps days remaining (variable "numDays") to the array in the translation json? And how does it select the right json-file with translations depending on language settings?

I sat down and tried to have a look at it yesterday, and I more or less came to the same conclusion. No errors in the console or Home Assistants's error log. Can't see where "the translation" is done either.

Tried to look at what was released in 2023.4.x and don't think I can see what could be affecting this issue. There is a chapter in 2023.4.0 regarding [Translating entities ](https://www.home-assistant.io/blog/2023/04/05/release-20234/#translating-entities), but there isn't much explained.
And in the 2023.4.1->2023.4.4 the most noteworthy I can find is the bumping of the frontend, but don't know where to find exactly what was done.

Since the number is fetched correctly, everything seems to be working except the "translation", so the problem is pretty much narrowed down

**Comment by angarok** [comment]
> > Have been taking a look at the code to try to figure out what stopped working. I'm not a coder so I can't really do much here. One observation is that the natural language strings are actually included in the custom component, not generated by HA Core. But I can't find where in the code the sensor maps days remaining (variable "numDays") to the array in the translation json? And how does it select the right json-file with translations depending on language settings?
> 
> I sat down and tried to have a look at it yesterday, and I more or less came to the same conclusion. No errors in the console or Home Assistants's error log. Can't see where "the translation" is done either.
> 
> Tried to look at what was released in 2023.4.x and don't think I can see what could be affecting this issue. There is a chapter in 2023.4.0 regarding [Translating entities ](https://www.home-assistant.io/blog/2023/04/05/release-20234/#translating-entities), but there isn't much explained. And in the 2023.4.1->2023.4.4 the most noteworthy I can find is the bumping of the frontend, but don't know where to find exactly what was done.
> 
> Since the number is fetched correctly, everything seems to be working except the "translation", so the problem is pretty much narrowed down

I have also been trying to figure this out. Reading the documentation it seems that is done correctly by @DSorlov, the component translation works but not that of the sensor states. My guess is that something has changed after 2023.4 but the documentation is not updated to reflect the changes.

**Comment by droidgren** [comment]
Please fix this, everybody has this issue.

**Comment by DSorlov** [comment]
@droidgren as soon as time permits or if someone have time to submit a PR, I dont have a permanent lab and I cannot replicate this in my own install..  :)

**Comment by angarok** [comment]
I looked at the documentation and it has been updated, it was actually two weeks ago. The translation files in the swemail integration needs to be reformatted to conform with the new translation method indroduced in 2023.4. I will try to take a look at it but as I said earlier I am not a coder, so no promises.

On another note, I invite everyone to keep comments in the thread focused on solving the issue, just asking (demanding…) for it to be fixed is not productive.

**Comment by JanSund** [comment]
Is this the documentation of changes i HA? https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/

**Comment by angarok** [comment]
> Is this the documentation of changes i HA? https://developers.home-assistant.io/blog/2023/03/27/entity_name_translations/

Yes, this is the blog post detailing how translations need to be implemented after 2023.4, it links to the actual documentation pages in the bottom. The documentation that was not updated before two+ weeks ago is the one detailing how custom integrations should handle translations.

https://developers.home-assistant.io/docs/internationalization/custom_integration

This now states that all translations need to be in a single file named with the corresponding language code, e.g. sv.json. Before you could just include a separate file for the entity domain you wanted to provide translations for, e.g sensor.sv.json, but this is no longer supported.

It also says that the translation files need to be like all other core integrations. This means that to translate a state of an entity it has to specify a translation key that maps to the translated string in the translation file. 

So to sum up everything that needs to be done:

1. The sensor declaration needs to be updated to include a translation_key.
2. The translation file needs to include the state translations under that translation_key.

This is a bit more than just copying the information from sensor.xx.json into xx.json. Still manageable, but will take more time.

**Comment by angarok** [comment]
Ok, so I solved it. Was easier than I thought. :-). I have it running locally with proper function. Now I just need to figure out how to make a PR, never done that before. Maybe later today or tomorrow, have life stuff that needs my time for now.

**Comment by thulin82** [comment]
> Ok, so I solved it. Was easier than I thought. :-). I have it running locally with proper function. Now I just need to figure out how to make a PR, never done that before. Maybe later today or tomorrow, have life stuff that needs my time for now.

Let me know if tou need help otherwise there should be a million guides out there. 
Fork - Branch - Commit - Push - PR

**Comment by angarok** [comment]
PR #20 submitted that fixes the issue

**Comment by nikee73** [comment]
Still no update in HA

**Comment by RH-Dreambox** [comment]
You can also edit existing files using the attached link.
That's what I did and it works great.
[https://github.com/DSorlov/swemail/pull/20/files](url)
